File: auth/internal/domain/service/user_service.go
package service

import (
	"context"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"

	"microservices/auth/internal/domain/entity"
	"microservices/auth/internal/domain/repository"
	"microservices/auth/internal/infrastructure/cache"
	"microservices/auth/pkg/errors"
	"microservices/auth/pkg/logger"
)

const (
	PermissionsTTL        = time.Hour
	MaxFailedAttempts     = 5
	LockDurationMinutes   = 30
)

type UserResponse struct {
	ID        uuid.UUID `json:"id"`
	Email     string    `json:"email"`
	FullName  string    `json:"fullName"`
	IsActive  bool      `json:"isActive"`
	Roles     []string  `json:"roles"`
	CreatedAt time.Time `json:"createdAt"`
	UpdatedAt time.Time `json:"updatedAt"`
}

type CreateUserInput struct {
	Email    string
	Password string
	FullName string
}

type UserService struct {
	userRepo           repository.UserRepository
	roleRepo           repository.RoleRepository
	userRoleRepo       repository.UserRoleRepository
	rolePermissionRepo repository.RolePermissionRepository
	loginHistoryRepo   repository.LoginHistoryRepository
	cache              *cache.RedisClient
}

func NewUserService(
	userRepo repository.UserRepository,
	roleRepo repository.RoleRepository,
	userRoleRepo repository.UserRoleRepository,
	rolePermissionRepo repository.RolePermissionRepository,
	loginHistoryRepo repository.LoginHistoryRepository,
	cache *cache.RedisClient,
) *UserService {
	return &UserService{
		userRepo:           userRepo,
		roleRepo:           roleRepo,
		userRoleRepo:       userRoleRepo,
		rolePermissionRepo: rolePermissionRepo,
		loginHistoryRepo:   loginHistoryRepo,
		cache:              cache,
	}
}

func (s *UserService) CreateUser(ctx context.Context, input *CreateUserInput) (*entity.User, error) {
	// Check if email exists
	exists, err := s.userRepo.ExistsByEmail(ctx, input.Email)
	if err != nil {
		return nil, errors.Wrap(err, "DATABASE_ERROR", "Failed to check email existence", 500)
	}
	if exists {
		return nil, errors.ErrEmailExists
	}

	// Hash password
	passwordHash, err := HashPassword(input.Password)
	if err != nil {
		return nil, errors.Wrap(err, "HASH_ERROR", "Failed to hash password", 500)
	}

	// Create user
	user := &entity.User{
		Email:        input.Email,
		PasswordHash: passwordHash,
		FullName:     input.FullName,
		Status:       entity.UserStatusActive,
	}

	if err := s.userRepo.Create(ctx, user); err != nil {
		return nil, errors.Wrap(err, "DATABASE_ERROR", "Failed to create user", 500)
	}

	// Assign default role (USER)
	defaultRole, err := s.roleRepo.FindDefault(ctx)
	if err != nil {
		if err != gorm.ErrRecordNotFound {
			logger.Error().Err(err).Msg("Failed to find default role")
		}
		// Try to find USER role
		defaultRole, err = s.roleRepo.FindByName(ctx, entity.RoleUser)
		if err != nil {
			logger.Error().Err(err).Msg("Failed to find USER role")
		}
	}

	if defaultRole != nil {
		if err := s.userRoleRepo.AssignRole(ctx, user.ID, defaultRole.ID, nil); err != nil {
			logger.Error().Err(err).Msg("Failed to assign default role")
		}
	}

	return user, nil
}

func (s *UserService) FindByID(ctx context.Context, id uuid.UUID) (*entity.User, error) {
	user, err := s.userRepo.FindByID(ctx, id)
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, errors.ErrUserNotFound
		}
		return nil, errors.Wrap(err, "DATABASE_ERROR", "Failed to find user", 500)
	}
	return user, nil
}

func (s *UserService) FindByEmail(ctx context.Context, email string) (*entity.User, error) {
	user, err := s.userRepo.FindByEmail(ctx, email)
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, errors.ErrUserNotFound
		}
		return nil, errors.Wrap(err, "DATABASE_ERROR", "Failed to find user", 500)
	}
	return user, nil
}

func (s *UserService) ValidateCredentials(ctx context.Context, email, password string) (*entity.User, error) {
	user, err := s.userRepo.FindByEmailWithPassword(ctx, email)
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, errors.ErrInvalidCredentials
		}
		return nil, errors.Wrap(err, "DATABASE_ERROR", "Failed to find user", 500)
	}

	// Check if account is locked
	if user.IsLocked() {
		return nil, errors.ErrAccountLocked
	}

	// Check user status
	switch user.Status {
	case entity.UserStatusInactive:
		return nil, errors.ErrUserInactive
	case entity.UserStatusSuspended:
		return nil, errors.ErrUserSuspended
	case entity.UserStatusBanned:
		return nil, errors.ErrUserBanned
	}

	// Verify password
	if !CheckPassword(password, user.PasswordHash) {
		// Increment failed attempts
		if err := s.userRepo.IncrementFailedAttempts(ctx, user.ID); err != nil {
			logger.Error().Err(err).Msg("Failed to increment failed attempts")
		}

		// Check if account should be locked
		if user.FailedLoginAttempts+1 >= MaxFailedAttempts {
			lockUntil := time.Now().Add(LockDurationMinutes * time.Minute)
			if err := s.userRepo.LockAccount(ctx, user.ID, lockUntil); err != nil {
				logger.Error().Err(err).Msg("Failed to lock account")
			}
		}

		return nil, errors.ErrInvalidCredentials
	}

	return user, nil
}

func (s *UserService) GetUserWithRoles(ctx context.Context, userID uuid.UUID) (*entity.User, error) {
	user, err := s.userRepo.GetUserWithRoles(ctx, userID)
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return nil, errors.ErrUserNotFound
		}
		return nil, errors.Wrap(err, "DATABASE_ERROR", "Failed to get user with roles", 500)
	}
	return user, nil
}

func (s *UserService) GetUserPermissions(ctx context.Context, userID uuid.UUID) ([]string, error) {
	return s.rolePermissionRepo.GetUserPermissions(ctx, userID)
}

func (s *UserService) GetUserPermissionsCached(ctx context.Context, userID uuid.UUID) ([]string, error) {
	// Try cache first
	permissions, err := s.cache.GetUserPermissions(ctx, userID.String())
	if err == nil && len(permissions) > 0 {
		return permissions, nil
	}

	// Fetch from database
	permissions, err = s.GetUserPermissions(ctx, userID)
	if err != nil {
		return nil, err
	}

	// Cache permissions
	if len(permissions) > 0 {
		if err := s.cache.CacheUserPermissions(ctx, userID.String(), permissions, PermissionsTTL); err != nil {
			logger.Warn().Err(err).Msg("Failed to cache user permissions")
		}
	}

	return permissions, nil
}

func (s *UserService) InvalidateUserPermissionsCache(ctx context.Context, userID uuid.UUID) error {
	return s.cache.InvalidateUserPermissions(ctx, userID.String())
}

func (s *UserService) HasPermission(ctx context.Context, userID uuid.UUID, permission string) (bool, error) {
	permissions, err := s.GetUserPermissionsCached(ctx, userID)
	if err != nil {
		return false, err
	}

	// Check for wildcard (super admin)
	for _, p := range permissions {
		if p == "*" {
			return true, nil
		}
		if p == permission {
			return true, nil
		}
	}

	return false, nil
}

func (s *UserService) HasAnyPermission(ctx context.Context, userID uuid.UUID, requiredPermissions []string) (bool, error) {
	permissions, err := s.GetUserPermissionsCached(ctx, userID)
	if err != nil {
		return false, err
	}

	permissionSet := make(map[string]bool)
	for _, p := range permissions {
		if p == "*" {
			return true, nil
		}
		permissionSet[p] = true
	}

	for _, required := range requiredPermissions {
		if permissionSet[required] {
			return true, nil
		}
	}

	return false, nil
}

func (s *UserService) AssignRole(ctx context.Context, userID uuid.UUID, roleName string, grantedBy *uuid.UUID) error {
	role, err := s.roleRepo.FindByName(ctx, roleName)
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return errors.ErrRoleNotFound
		}
		return errors.Wrap(err, "DATABASE_ERROR", "Failed to find role", 500)
	}

	// Check if role already assigned
	exists, err := s.userRoleRepo.Exists(ctx, userID, role.ID)
	if err != nil {
		return errors.Wrap(err, "DATABASE_ERROR", "Failed to check role assignment", 500)
	}
	if exists {
		return nil // Already assigned
	}

	if err := s.userRoleRepo.AssignRole(ctx, userID, role.ID, grantedBy); err != nil {
		return errors.Wrap(err, "DATABASE_ERROR", "Failed to assign role", 500)
	}

	// Invalidate permissions cache
	if err := s.InvalidateUserPermissionsCache(ctx, userID); err != nil {
		logger.Warn().Err(err).Msg("Failed to invalidate permissions cache")
	}

	return nil
}

func (s *UserService) RemoveRole(ctx context.Context, userID uuid.UUID, roleName string) error {
	role, err := s.roleRepo.FindByName(ctx, roleName)
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			return errors.ErrRoleNotFound
		}
		return errors.Wrap(err, "DATABASE_ERROR", "Failed to find role", 500)
	}

	if err := s.userRoleRepo.Delete(ctx, userID, role.ID); err != nil {
		return errors.Wrap(err, "DATABASE_ERROR", "Failed to remove role", 500)
	}

	// Invalidate permissions cache
	if err := s.InvalidateUserPermissionsCache(ctx, userID); err != nil {
		logger.Warn().Err(err).Msg("Failed to invalidate permissions cache")
	}

	return nil
}

func (s *UserService) GetUserRoles(ctx context.Context, userID uuid.UUID) ([]string, error) {
	userRoles, err := s.userRoleRepo.FindByUserIDWithRoles(ctx, userID)
	if err != nil {
		return nil, errors.Wrap(err, "DATABASE_ERROR", "Failed to get user roles", 500)
	}

	roles := make([]string, 0, len(userRoles))
	for _, ur := range userRoles {
		if !ur.IsExpired() {
			roles = append(roles, ur.Role.Name)
		}
	}

	return roles, nil
}

func (s *UserService) UpdateLoginInfo(ctx context.Context, userID uuid.UUID, ip string) error {
	return s.userRepo.IncrementLoginCount(ctx, userID, ip)
}

func (s *UserService) RecordLoginHistory(ctx context.Context, userID uuid.UUID, status entity.LoginStatus, authMethod entity.AuthMethod, ipAddress string, userAgent *string) error {
	history := entity.NewLoginHistory(userID, status, authMethod, ipAddress, userAgent)
	return s.loginHistoryRepo.Create(ctx, history)
}

func (s *UserService) ToUserResponse(user *entity.User) *UserResponse {
	return &UserResponse{
		ID:        user.ID,
		Email:     user.Email,
		FullName:  user.FullName,
		IsActive:  user.IsActive(),
		Roles:     user.GetRoleNames(),
		CreatedAt: user.CreatedAt,
		UpdatedAt: user.UpdatedAt,
	}
}

File: auth-profile-nestjs/auth/src/modules/identity/services/user.service.ts
import {
  Injectable,
  ConflictException,
  NotFoundException,
  InternalServerErrorException,
  Logger,
} from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import * as bcrypt from 'bcrypt';

import { User, Role, UserRole } from '../entities';
import { UserResponse } from '../interfaces';
import { TOKEN_CONFIG, ERROR_MESSAGES, DEFAULT_ROLES } from '../constants';
import { RegisterDto } from '../dto';
import { RedisCacheService } from './redis-cache.service';

@Injectable()
export class UserService {
  private readonly logger = new Logger(UserService.name);

  constructor(
    @InjectRepository(User)
    private readonly userRepository: Repository<User>,
    @InjectRepository(Role)
    private readonly roleRepository: Repository<Role>,
    @InjectRepository(UserRole)
    private readonly userRoleRepository: Repository<UserRole>,
    private readonly redisCacheService: RedisCacheService,
  ) {}

  /**
   * Create a new user
   */
  async createUser(dto: RegisterDto): Promise<User> {
    try {
      // Check if email already exists
      const existingUser = await this.userRepository.findOne({
        where: { email: dto.email.toLowerCase() },
      });

      if (existingUser) {
        throw new ConflictException(ERROR_MESSAGES.EMAIL_EXISTS);
      }

      // Hash password
      const passwordHash = await bcrypt.hash(dto.password, TOKEN_CONFIG.BCRYPT_SALT_ROUNDS);

      // Create user entity
      const user = this.userRepository.create({
        email: dto.email.toLowerCase(),
        passwordHash,
        fullName: dto.fullName,
        status: 'ACTIVE',
      });

      const savedUser = await this.userRepository.save(user);

      // Assign default role 'CUSTOMER'
      await this.assignRole(savedUser.id, DEFAULT_ROLES.CUSTOMER);

      return savedUser;
    } catch (error) {
      if (error instanceof ConflictException) {
        throw error;
      }
      this.logger.error(
        'Failed to create user',
        error instanceof Error ? error.stack : String(error),
      );
      throw new InternalServerErrorException(ERROR_MESSAGES.REGISTRATION_FAILED);
    }
  }

  /**
   * Find user by ID
   */
  async findById(id: string): Promise<User | null> {
    return this.userRepository.findOne({
      where: { id, status: 'ACTIVE' },
    });
  }

  /**
   * Find user by email with password hash (for authentication)
   */
  async findByEmailWithPassword(email: string): Promise<User | null> {
    return this.userRepository
      .createQueryBuilder('user')
      .addSelect('user.passwordHash')
      .where('user.email = :email', { email: email.toLowerCase() })
      .andWhere('user.status = :status', { status: 'ACTIVE' })
      .getOne();
  }

  /**
   * Validate user credentials
   */
  async validateCredentials(email: string, password: string): Promise<User | null> {
    const user = await this.findByEmailWithPassword(email);

    if (!user) {
      return null;
    }

    const isPasswordValid = await bcrypt.compare(password, user.passwordHash);

    if (!isPasswordValid) {
      return null;
    }

    return user;
  }

  /**
   * Get user with roles
   */
  async getUserWithRoles(userId: string): Promise<User | null> {
    return this.userRepository.findOne({
      where: { id: userId, status: 'ACTIVE' },
      relations: ['userRoles', 'userRoles.role'],
    });
  }

  /**
   * Get user permissions from database
   */
  async getUserPermissions(userId: string): Promise<string[]> {
    try {
      const userRoles = await this.userRoleRepository.find({
        where: { userId },
        relations: ['role', 'role.rolePermissions', 'role.rolePermissions.permission'],
      });

      const permissions = new Set<string>();

      for (const userRole of userRoles) {
        // Check for super admin (all permissions)
        if (userRole.role?.name === DEFAULT_ROLES.SUPER_ADMIN) {
          permissions.add('*');
        }

        if (userRole.role?.rolePermissions) {
          for (const rolePermission of userRole.role.rolePermissions) {
            if (rolePermission.permission?.slug) {
              permissions.add(rolePermission.permission.slug);
            }
          }
        }
      }

      return Array.from(permissions);
    } catch (error) {
      this.logger.error(
        `Failed to fetch permissions for user ${userId}`,
        error instanceof Error ? error.stack : String(error),
      );
      return [];
    }
  }

  /**
   * Get user permissions (with caching)
   */
  async getUserPermissionsCached(userId: string): Promise<string[]> {
    // Try cache first
    let permissions = await this.redisCacheService.getUserPermissions(userId);

    if (!permissions) {
      // Cache miss - fetch from DB and cache
      permissions = await this.getUserPermissions(userId);
      await this.redisCacheService.cacheUserPermissions(userId, permissions);
    }

    return permissions;
  }

  /**
   * Assign a role to user
   */
  async assignRole(userId: string, roleName: string): Promise<void> {
    try {
      const role = await this.roleRepository.findOne({
        where: { name: roleName },
      });

      if (!role) {
        this.logger.warn(`Role ${roleName} not found`);
        return;
      }

      // Check if already assigned
      const existing = await this.userRoleRepository.findOne({
        where: { userId, roleId: role.id },
      });

      if (existing) {
        return;
      }

      const userRole = this.userRoleRepository.create({
        userId,
        roleId: role.id,
      });

      await this.userRoleRepository.save(userRole);

      // Invalidate permissions cache
      await this.redisCacheService.invalidateUserPermissions(userId);
    } catch (error) {
      this.logger.error(
        `Failed to assign role ${roleName} to user ${userId}`,
        error instanceof Error ? error.stack : String(error),
      );
    }
  }

  /**
   * Remove a role from user
   */
  async removeRole(userId: string, roleName: string): Promise<void> {
    try {
      const role = await this.roleRepository.findOne({
        where: { name: roleName },
      });

      if (!role) {
        return;
      }

      await this.userRoleRepository.delete({ userId, roleId: role.id });

      // Invalidate permissions cache
      await this.redisCacheService.invalidateUserPermissions(userId);
    } catch (error) {
      this.logger.error(
        `Failed to remove role ${roleName} from user ${userId}`,
        error instanceof Error ? error.stack : String(error),
      );
    }
  }

  /**
   * Get user roles as string array
   */
  async getUserRoles(userId: string): Promise<string[]> {
    const userRoles = await this.userRoleRepository.find({
      where: { userId },
      relations: ['role'],
    });

    return userRoles.map((ur) => ur.role.name);
  }

  /**
   * Update user profile
   */
  async updateProfile(userId: string, updates: { fullName?: string }): Promise<User> {
    const user = await this.findById(userId);

    if (!user) {
      throw new NotFoundException(ERROR_MESSAGES.USER_NOT_FOUND);
    }

    if (updates.fullName) {
      user.fullName = updates.fullName;
    }

    return this.userRepository.save(user);
  }

  /**
   * Deactivate user
   */
  async deactivateUser(userId: string): Promise<void> {
    await this.userRepository.update({ id: userId }, { status: 'INACTIVE' });
    await this.redisCacheService.invalidateUserPermissions(userId);
  }

  /**
   * Transform User entity to UserResponse
   */
  async toUserResponse(user: User): Promise<UserResponse> {
    const roles = await this.getUserRoles(user.id);

    return {
      id: user.id,
      email: user.email,
      fullName: user.fullName,
      isActive: user.status === 'ACTIVE',
      roles,
      createdAt: user.createdAt,
      updatedAt: user.updatedAt,
    };
  }

  /**
   * Check if user has specific permission
   */
  async hasPermission(userId: string, permission: string): Promise<boolean> {
    const permissions = await this.getUserPermissionsCached(userId);

    // Super admin has all permissions
    if (permissions.includes('*')) {
      return true;
    }

    return permissions.includes(permission);
  }

  /**
   * Check if user has any of the specified permissions
   */
  async hasAnyPermission(userId: string, requiredPermissions: string[]): Promise<boolean> {
    const permissions = await this.getUserPermissionsCached(userId);

    // Super admin has all permissions
    if (permissions.includes('*')) {
      return true;
    }

    return requiredPermissions.some((p) => permissions.includes(p));
  }
}
